### PR TITLE
[REFACTOR] 회원 탈퇴 시 관련 엔티티 삭제 

### PR DESCRIPTION
**Pull request by 5hseok**
## #️⃣ Related Issue

Linked Feature backlog: #82 

## 📝 Purpose of PR

회원 탈퇴 시 회원에 연관된 엔티티들을 삭제

## Contents

탈퇴 시 회원과 관련된 엔티티를 삭제하는 방향으로 정책 변경
statistics는 BaseEntity를 상속하지 않기에 soft delete 시키지 않고 바로 삭제시키고, 
다른 테이블들은 deletedAt이 존재하기에 soft delete 처리

User, UserAnswer은 User 도메인에서 처리하고 나머지 엔티티들은 Event를 발행하여 이벤트 기반으로 동작 

## Checklist

-   [ ] Does commit messages follow the convention?
-   [ ] Are variable names brief but descriptive?
-   [ ] Is the code clean and easy to understand?
-   [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
-   [ ] Does the test successfully test edge and corner cases?

## 💬 Review Requirements (Optional)

Is there anything you can praise about this PR? Start with praise.
Make a review, leaving kind comments and suggesting changes where needed (to resolve the above).

### CODE REVIEW

File: src/test/java/ac/dnd/dodal/application/plan_feedback/service/PlanFeedbackEventListenerTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+public class PlanFeedbackEventListenerTest extends IntegrationTest {
+
+    @Autowired
+    private ApplicationEventPublisher eventPublisher;
+
+    @Autowired
+    private PlanFeedbackService planFeedbackService;
+
+    @Test
+    @DisplayName("계획이 삭제되면 해당 계획과 관련된 모든 계획 피드백이 삭제된다.")
```
Comment (5hseok) [inline src/test/java/ac/dnd/dodal/application/plan_feedback/service/PlanFeedbackEventListenerTest.java:26]: 이 부분에서는 DeletedPlanEvent를 받아서 처리하기 때문에 PlanCommandServiceTest에서 plan을 delete하는 부분과 겹치더라고요?

번갈아가면서 테스트가 1개 실패, 전체 성공을 반복하는 상황입니다...

**Comment by 5hseok** [comment]
**User 탈퇴 이벤트 흐름** 

1. User, UserAnswer 삭제 후 UserWithdrawn 이벤트 발행 

	-> UserWithdrawn 이벤트를 받아 UserGuide, Goal 삭제 

2. Goal 삭제 후 DeletedGoalEvent 발행

	-> DeletedGoalEvent 를 받아 GoalStatistics, PlanHistory, Plan 삭제 

3. Plan 삭제 후 DeletedPlanEvent 발행 

	-> DeletedPlanEvent 를 받아 PlanFeedback, HistoryStatistics 삭제

**Comment by 5hseok** [comment]
도메인마다의 삭제 로직은 거의 비슷합니다

File: src/test/java/ac/dnd/dodal/application/plan_feedback/service/PlanFeedbackEventListenerTest.java (hunk 1 above)
Comment (sueelly) [reply src/test/java/ac/dnd/dodal/application/plan_feedback/service/PlanFeedbackEventListenerTest.java:26]: 테스트를 독립적으로 분리해야할거같아요
table truncate로직을 추가하고 해결하는 수밖에...

File: src/test/java/ac/dnd/dodal/domain/user/UserFixture.java [hunk 2]
```diff
@@ -19,7 +20,7 @@ public static OAuthUserInfoRequestDto createRequest(String appleId) {
     return new OAuthUserInfoRequestDto(NICKNAME, EMAIL, profileImage, appleId, deviceToken);
   }
 
-  public static User createUser(String email, String nickname, String profileImageUrl, String deviceToken) {
```
Comment (sueelly) [inline src/test/java/ac/dnd/dodal/domain/user/UserFixture.java:22]: 기본 생성 메서드 그대로 두고, 이걸 활용해서 createUser()로 default 유저 생성하도록 하는게 확장성에 있어서 더 좋을 것 같아요~

File: src/main/java/ac/dnd/dodal/common/util/OAuth2Util.java [hunk 3]
```diff
@@ -180,7 +180,7 @@ public <T> T decodePayload(String token, Class<T> targetClass) {
     }
   }
 
-  public void revokeAppleToken(String accessToken) {
+  public boolean revokeAppleToken(String accessToken) {
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/common/util/OAuth2Util.java:183]: 혹시 boolean으로 바뀐 이유가 뭘까요? 실제로 값을 사용하는 부분도 안보여서요

File: src/main/java/ac/dnd/dodal/application/plan_feedback/repository/PlanFeedbackRepository.java [hunk 4]
```diff
@@ -1,11 +1,17 @@
...
 
 import ac.dnd.dodal.domain.plan_feedback.model.PlanFeedback;
 import ac.dnd.dodal.domain.plan_feedback.model.PlanFeedbackId;
 
+import java.util.List;
+
 @Repository
 public interface PlanFeedbackRepository extends JpaRepository<PlanFeedback, PlanFeedbackId> {
+
+    @Query("SELECT pf FROM plan_feedbacks pf WHERE pf.plan.planId = :planId")
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/application/plan_feedback/repository/PlanFeedbackRepository.java:15]: 전체 조회할 때 deletedAt is NULL의 조건이 붙는 곳이 있고 아닌 곳이 있는데, 통일하면 좋을 것 같아요
다만 저도 도메인 로직에서 통일되게 처리하지 못해서 deleted된 것에 다시 delete 수행 시 예외가 발생할 수도 있고, 그냥 deletedAt이 업데이트만 될수도 있어요

**Review by sueelly — CHANGES_REQUESTED** [review]
전체적인 플로우 좋아요~ 다만 몇 가지 수정사항이 있는데

1. 전체 조회 시 deletedAt is NULL 조건 사용 통일하기
2. 이벤트의 경우 아무 옵션을 설정하지 않으면 동기적으로, 하나의 트랜잭션 내에서 수행돼요. 근데 모든 것을 한번에 삭제하도록 하면 응답시간이 많이 지연되고, 많은 쓰기가 한번에 이루어져 db에도 부담이 될 것 같아요. 비동기적으로 처리하도록 바꾸는 게 좋을 것 같아요
3. withdrawn acceptance test에서 withdraw 이후 user 조회 api 호출 후 제대로 삭제됐는지 확인하는 로직 들어가면 좋을 것 같아요~ 말 그대로 시나리오 테스트니까요

File: src/main/java/ac/dnd/dodal/common/util/OAuth2Util.java (hunk 3 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/common/util/OAuth2Util.java:183]: 추후에 apple 탈퇴 테스트를 처리할 때 검증하기 위해 void에서 boolean으로 수정했습니다.

File: src/main/java/ac/dnd/dodal/application/plan_feedback/repository/PlanFeedbackRepository.java (hunk 4 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/application/plan_feedback/repository/PlanFeedbackRepository.java:15]: 찾아보고 deletedAt 조건이 없는 부분에 추가해두는 게 좋을 것 같아요. 한번에 코드 수정하겠습니다.

**Review by sueelly — APPROVED** [review]
good~ 
비동기 실패 처리 로직은 추후 생각해보시죵
